### PR TITLE
refactored ledger approvement API

### PR DESCRIPTION
**Pull request by AndyGura**

### CODE REVIEW

File: src/ledger/ledger.mo [hunk 1]
```diff
@@ -354,33 +367,31 @@ module {
...
         case (#err error) { return #err(error) };
         case (_) {}
       };
-      let txSize = Tx.size(tx);
-      tracker.recordTx(#submit);
-      let approvals: MutableApprovals = Array.tabulateVar(tx.map.size(), func (i: Nat): Bool = tx.map[i].owner == caller or (tx.map[i].outflow.size() == 0 and tx.map[i].mints.size() == 0 and tx.map[i].burns.size() == 0));
+      let approvals: MutableApprovals = buildApprovals(caller, tx);
+      if (Array.foldRight(Array.freeze(approvals), true, Bool.logand)) {
+        return #err(#AlreadyApproved);
+      };
```
Comment (timohanke) [inline src/ledger/ledger.mo:382]: We can remove this and #AlreadyApproved can be removed from SubmitResult.

If it is fully approved then fine. Someone can call execute on it afterwards.

File: src/ledger/ledger.mo [hunk 2]
```diff
@@ -33,14 +33,12 @@ module {
...
   The vector's length equals the number of contributors to the transactions.
-  When a txreq is fully approved then the ledger processes it.
   */
   public type TxReq = {
     tx : Tx.Tx;
     submitter : Principal;
     var lid : ?LocalId;
-    var status: { #unapproved : MutableApprovals; #rejected; #processed; #pending; #failed: ProcessingError; };
-    size: Nat;
+    var status: { #unapproved : MutableApprovals; #rejected; #processed; #failed: ProcessingError; };
```
Comment (timohanke) [inline src/ledger/ledger.mo:41]: #unapproved is not a good name anymore. We can have fully approved requests that are not yet executed. Maybe #waiting?

File: src/ledger/ledger.mo [hunk 3]
```diff
@@ -71,6 +69,9 @@ module {
     #InsufficientFunds;
     #NotAController;  // in attempted mint operation
   };
+  public type SubmitImmediateError = ProcessingError or { #NotApproved; };
```
Comment (timohanke) [inline src/ledger/ledger.mo:72]: Rename to SubmitAndExecuteError

File: src/ledger/ledger.mo (hunk 1 above)
Comment (AndyGura) [reply src/ledger/ledger.mo]: Done

File: src/ledger/ledger.mo (hunk 2 above)
Comment (AndyGura) [reply src/ledger/ledger.mo]: Renamed

File: src/ledger/ledger.mo (hunk 3 above)
Comment (AndyGura) [reply src/ledger/ledger.mo]: Done

File: src/ledger/ledger.mo [hunk 4]
```diff
@@ -49,15 +47,15 @@ module {
     tx : Tx.Tx;
     submitter : Principal;
     gid : GlobalId;
-    status : { #unapproved : Approvals; #rejected; #processed; #pending; #failed: ProcessingError; };
+    status : { #waiting : Approvals; #rejected; #processed; #failed: ProcessingError; };
```
Comment (timohanke) [inline src/ledger/ledger.mo:50]: Sorry, maybe #pending is a better name. I first thought that "pending" was used for something different in the aggregator. But now seeing that we have #NotPendingError that seems to match. A #NotPendingError occurs if the request status is _not_ #waiting, or? If that's the case then we should use the same name in the status as in the error.

Comment (AndyGura) [reply src/ledger/ledger.mo:50]: Right, #pending sounds more appropriate, but can be confused with status #pending in the aggregator, there it means "sent to the ledger, no response yet". Should we rename this status in aggregator to something else?

Comment (AndyGura) [reply src/ledger/ledger.mo]: Renamed to #pending

Comment (AndyGura) [reply src/ledger/ledger.mo]: Done

File: src/ledger/ledger_api.mo [hunk 5]
```diff
@@ -63,30 +63,39 @@ actor class LedgerAPI(initialAggregators : [Principal], lookupTableCapacity: Nat
     };
   };
 
-  /** Create a new transaction request.
+  /** Create a new transaction request, which requires approvement by one or more contributors
```
Comment (timohanke) [inline src/ledger/ledger_api.mo:66]: approval (instead of approvement)

**Review by timohanke — APPROVED** [review]

File: src/ledger/ledger_api.mo (hunk 5 above)
Comment (AndyGura) [reply src/ledger/ledger_api.mo:66]: fixed

**Pull request merged**